src/gui/properties/speedplotview.cpp

```diff
@@ -118,55 +107,81 @@ namespace
     }
 }
 
-SpeedPlotView::Averager::Averager(int divider, boost::circular_buffer<PointData> &sink)
-    : m_divider(divider)
-    , m_sink(sink)
-    , m_counter(0)
-    , m_accumulator {}
+SpeedPlotView::Averager::Averager(const milliseconds duration, const milliseconds resolution)
+    : m_resolution {resolution}
+    , m_maxDuration {duration}
 {
+    m_lastSampleTime.start();
 }
 
-void SpeedPlotView::Averager::push(const PointData &pointData)
+bool SpeedPlotView::Averager::push(const SampleData &sampleData)
 {
     // Accumulator overflow will be hit in worst case on longest used averaging span,
     // defined by divider value. Maximum divider is DIVIDER_24HOUR = 144
     // Using int32 for accumulator we get overflow when transfer speed reaches 2^31/144 ~~ 14.2 MBytes/s.
     // With quint64 this speed limit is 2^64/144 ~~ 114 PBytes/s.
     // This speed is inaccessible to an ordinary user.
-    m_accumulator.x += pointData.x;
+    ++m_counter;
     for (int id = UP; id < NB_GRAPHS; ++id)
-        m_accumulator.y[id] += pointData.y[id];
-    m_counter = (m_counter + 1) % m_divider;
-    if (m_counter != 0)
-        return; // still accumulating
+        m_accumulator[id] += sampleData[id];
+
+    // system may go to sleep, that can cause very big elapsed interval
+    const milliseconds updateInterval {static_cast<int64_t>(BitTorrent::Session::instance()->refreshInterval() * 1.25)};
+    const milliseconds maxElapsed {std::max(updateInterval, m_resolution)};
+    const milliseconds elapsed {std::min(milliseconds {m_lastSampleTime.elapsed()}, maxElapsed)};
+    if (elapsed < m_resolution)
+        return false; // still accumulating
+
     // it is time final averaging calculations
     for (int id = UP; id < NB_GRAPHS; ++id)
-        m_accumulator.y[id] /= m_divider;
-    m_accumulator.x /= m_divider;
+        m_accumulator[id] /= m_counter;
+
+    m_currentDuration += elapsed;
+    // remove extra data from front if we reached max duration
+    if (m_currentDuration > m_maxDuration)
+    {
+        // once we go above the max duration never go below that
+        // otherwise it will cause empty space in graphs
+        while(!m_sink.empty()
+               && ((m_currentDuration - m_sink.front().duration) > m_maxDuration))
+        {
+            m_currentDuration -= m_sink.front().duration;
+            m_sink.pop_front();
+        }
+    }
+
+    // we manually clear the extra data in above `if`, so there is always new data to `push`
+    if (m_sink.capacity() == m_sink.size())
+    {
+        const milliseconds totalDurationOfSink {m_currentDuration - elapsed}; // we haven't pushed yet
+        Q_ASSERT(totalDurationOfSink >= milliseconds {0});
+
+        m_sink.set_capacity(m_sink.size() + std::max<int64_t>((m_maxDuration - totalDurationOfSink) / elapsed, 1));
```

Comment:
The one thing I dislike in `Averager::push()` is playing with buffer capacity. I would make it static and pop/push items manually.

Reply:
> The one thing I dislike in Averager::push() is playing with buffer capacity. I would make it static and pop/push items manually.

IDK how that can be achieved, only push has the latest information to correctly set the buffer size

Reply:
You can calculate max buffer size using duration and resolution (aka duration/resolution) and set capacity to it. Since duration of each sample can't be less than resolution you will never overflow the buffer (if you manually pop outdated samples and then push new one like you do currently).